controllers/appgroup_controller.go

```diff
@@ -499,32 +506,66 @@ func (r *ApplicationGroupReconciler) cleanupWorkflow(ctx context.Context, logr l
 	_ = r.List(ctx, &wfs, listOption)
 
 	if wfs.Items.Len() != 0 {
-		logr.Info("Reversing the workflow")
-
 		wf := wfs.Items[0]
 		for _, node := range wf.Status.Nodes {
 			nodes[node.ID] = node
 		}
-		graph, err := workflow.Build(g.Name, nodes)
+		rwf := &v1alpha12.Workflow{}
+
+		rwfName := fmt.Sprintf("%s-reverse", wf.Name)
+		rwfNamespace := wf.Namespace
+		err := r.Client.Get(ctx, types.NamespacedName{Namespace: rwfNamespace, Name: rwfName}, rwf)
 		if err != nil {
-			logr.Error(err, "failed to build the wf status DAG")
-		}
+			if kerrors.IsNotFound(err) {
+				logr.Info("Reversing the workflow")
+
+				err = r.generateReverseWorkflow(ctx, logr, nodes, &wf)
+				if err != nil {
+					logr.Error(err, "failed to generate reverse workflow")
+					// if generation of reverse workflow failed, delete the forward workflow and return
+					err = r.Client.Delete(ctx, &wf)
+					if err != nil {
+						logr.Error(err, "failed to delete workflow CRO")
+						return false
+					}
+					return false
+				}
 
-		rev := graph.Reverse()
+				// reverse workflow started - requeue
+				return true
+			}
+			logr.Error(err, "failed to GET workflow object with an unrecoverable error")
+		} else {
+			// check the completion of the reverse workflow
+			if !rwf.Status.FinishedAt.IsZero() {
+				logr.Info("reverse workflow is finished")
 
-		for _, bucket := range rev {
-			for _, hr := range bucket {
-				err = r.Client.Delete(ctx, &hr)
+				err = r.Client.Delete(ctx, &wf)
```

Comment:
Don't we also need to delete the reverse workflow?

Reply:
Not needed since it is owned by the forward workflow using the ownersReference field. So it will get garbage collected on deleting the forward workflow object